collectors/scm/bitbucket/src/main/java/com/capitalone/dashboard/bitbucketapi/BitbucketUrlParser.java

```diff
@@ -0,0 +1,125 @@
+    private static final String SSH = "ssh";
+
+    /**
+     * This method constructs Bitbucket url parts from bitbucket url. Usefulness of this method is observed
+     * at the time of creating Bitbucket API endpoints.
+     *
+     * @param rawUrl
+     * @return BitbucketUrlParts parts of Bitbucket URL
+     */
+    public static BitbucketUrlParts parseBitbucketUrl(String rawUrl) {
+        String repoUrlProcessed = rawUrl;
+
+        if (repoUrlProcessed.endsWith(GIT_EXTENSION)) {
+            repoUrlProcessed =
+                    StringUtils.substring(
+                            repoUrlProcessed, 0, StringUtils.lastIndexOf(repoUrlProcessed, GIT_EXTENSION));
+        }
+
+        URI uri = URI.create(repoUrlProcessed.replaceAll(" ", "%20"));
+
+        String host = uri.getHost();
+        int port = uri.getPort();
+
+        String scheme = SSH.equalsIgnoreCase(uri.getScheme()) ? HTTPS : uri.getScheme();
+
+        String path = parseURLPath(uri);
+
+        String[] splitPath = path.split("/");
+        String projectKey;
+        String repositorySlug;
+
+        if (splitPath.length > 1) {
+            projectKey = splitPath[0];
+            repositorySlug = StringUtils.substring(path, StringUtils.indexOf(path, '/') + 1);
+        } else {
+            // Shouldn't get to this case
+            projectKey = "";
+            repositorySlug = path;
+        }
+
+        return new BitbucketUrlParts(scheme, host, port, projectKey, repositorySlug);
+    }
+
+    private static String parseURLPath(URI uri) {
+        String path = uri.getPath();
+        if ((path.startsWith(SCM_SUFFIXED_WITH_FORWARD_SLASH)
+                || path.startsWith(SCM_PREFIXED_WITH_FORWARD_SLASH))
+                && path.length() > 4) {
+            path = path.substring(4);
+        }
+        if (path.length() > 0 && path.charAt(0) == '/') {
+            path = path.substring(1);
+        }
+        return path;
+    }
+
+    /**
+     * Represents a Bitbucket URL constituents.
+     */
+    public static class BitbucketUrlParts {
+        private final String scheme;
+        private final String host;
+        private final Integer port;
+        private final String projectKey;
+        private final String repoKey;
+
+        public BitbucketUrlParts(
+                String scheme, String host, int port, String projectKey, String repositorySlug) {
+            this.scheme = scheme;
+            this.host = host;
+            this.port = port;
+            this.projectKey = projectKey;
+            this.repoKey = repositorySlug;
+        }
+
+        public String getScheme() {
+            return scheme;
+        }
+
+        public String getHost() {
+            return host;
+        }
+
+        public Integer getPort() {
+            return port;
+        }
+
+        public String getProjectKey() {
+            return projectKey;
+        }
+
+        public String getRepoKey() {
+            return repoKey;
+        }
+    }
+}
+
+/*
```

Comment:
These Copyright comments will need to be removed

Reply:
@Sbrenthughes These copyright comments were recommended to me in my company's internal OSS review. Is there no need to include any copyright section? Do you have any contribution guideline document which explains it?

Reply:
See https://github.com/Hygieia/Hygieia/blob/master/LICENSE

Reply:
@Sbrenthughes I saw that LICENSE. If you read the very last section of that license, It is recommended that developers should add the boilerplate copyright comment in the source file. What's the recommendation here?

-------
APPENDIX: How to apply the Apache License to your work.

      To apply the Apache License to your work, attach the following
      boilerplate notice, with the fields enclosed by brackets "{}"
      replaced with your own identifying information. (Don't include
      the brackets!)  The text should be enclosed in the appropriate
      comment syntax for the file format. We also recommend that a
      file or class name and description of purpose be included on the
      same "printed page" as the copyright notice for easier
      identification within third-party archives.

   Copyright [yyyy] [name of copyright owner]

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
--------

Reply:
Sure, let me get back to you on this.

Reply:
Thank you @Sbrenthughes for your help on this topic. Looking forward to your recommendation.

Reply:
@Sbrenthughes - Thanks for reaching out regarding the license header. We suggest the following header be added to the source files verbatim with only the [yyyy] text replaced with the year: 

**
SPDX-Copyright: Copyright (c) Capital One Services, LLC 

SPDX-License-Identifier: Apache-2.0 

Copyright [yyyy] Capital One Services, LLC 

 

Licensed under the Apache License, Version 2.0 (the "License"); 

you may not use this file except in compliance with the License. 

You may obtain a copy of the License at 

 

    http://www.apache.org/licenses/LICENSE-2.0 

 

Unless required by applicable law or agreed to in writing, software 

distributed under the License is distributed on an "AS IS" BASIS, 

WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 

See the License for the specific language governing permissions and limitations under the License. 

**

Reply:
@Sbrenthughes Just to confirm, you want me to add copyright headers(as mentioned by you). What about the copyright section I have in the file in the footer? Should I keep it?

Reply:
@Sbrenthughes Can you please confirm about the footer copyright section? Should I keep it?

Reply:
Yes remove footer. If you want you can use the header provided by Santhi. But the license file should cover the need to have these in each individual file.

Reply:
@Sbrenthughes I am seeking approval from my OSS team to remove these copyright. I will update you as soon as I get their response.

Reply:
From my understanding the below is what is left that needs to be done for the headers.
> @Sbrenthughes - Thanks for reaching out regarding the license header. We suggest the following header be added to the source files verbatim with only the [yyyy] text replaced with the year:
> 
> **
> SPDX-Copyright: Copyright (c) Capital One Services, LLC
> 
> SPDX-License-Identifier: Apache-2.0
> 
> Copyright [yyyy] Capital One Services, LLC
> 
> Licensed under the Apache License, Version 2.0 (the "License");
> 
> you may not use this file except in compliance with the License.
> 
> You may obtain a copy of the License at
> 
> ```
> http://www.apache.org/licenses/LICENSE-2.0 
> ```
> 
> Unless required by applicable law or agreed to in writing, software
> 
> distributed under the License is distributed on an "AS IS" BASIS,
> 
> WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> 
> See the License for the specific language governing permissions and limitations under the License.
> 
> **